### ISSUE TITLE
Fedora s390x abending when Miscellaneous-Instruction-Extensions Facility 3 is enabled

### ISSUE THREAD

**Issue by mcisho** [SDL-Hercules-390/hyperion#367]
I haven't used a Fedora s390x system under Hercules since mid-December 2020, when everything was fine. Last weekend I updated my Hercules to the (then) latest version. Today I ran a Fedora s390x system which program checked trying to execute Vector instructions. 

```
12:35:23 HHC01413I Hercules version 4.4.9999.0-SDL-ge6c80705 (4.4.9999.0)
12:35:23 HHC01414I (C) Copyright 1999-2020 by Roger Bowler, Jan Jaeger, and others
12:35:23 HHC01417I ** The SoftDevLabs version of Hercules **
12:35:23 HHC01415I Build date: Jan 23 2021 at 17:20:46
12:35:23 HHC01417I Built with: GCC 10.2.1 20201125 (Red Hat 10.2.1-9)
12:35:23 HHC01417I Build type: GNU/Linux x86_64 host architecture build
  ::
12:36:23 
12:36:23 Welcome to [0;38;2;60;110;180mFedora 33 (Thirty Three) dracut-050-64.git20200529
12:36:23 .fc33 (Initramfs)[0m!
12:36:23 
12:36:27 [   21.488279] User process fault: interruption code 0001 ilc:3 in libc-2.32.so[
12:36:27 3ff88980000+1c4000]
12:36:27 [   21.488420] CPU: 0 PID: 1 Comm: systemd Not tainted 5.9.13-200.fc33.s390x #1
12:36:27 [   21.488488] Hardware name: HRC 2817 EMULATOR EMULATOR (LPAR)
12:36:27 [   21.488551] User PSW : 0705000180000000 000003ff88a2f40a
12:36:27 [   21.488638]            R:0 T:1 IO:1 EX:1 Key:0 M:1 W:0 P:1 AS:0 CC:0 PM:0 RI:
12:36:27 0 EA:3
12:36:27 [   21.488728] User GPRS: 0000000000000000 000003ff88a2f3f0 000002aa144401f0 000
12:36:27 002aa144401f2
12:36:27 [   21.488800]            0000000000000001 fffffffffffffffe 000003ffd8a7e6a0 000
12:36:27 002aa14448e50
12:36:27 [   21.488879]            000002aa144401f0 000003ff89369820 0000000000000000 000
12:36:27 003ffd8a7e6a0
12:36:27 [   21.488955]            000003ff8932af70 000002aa144401d0 000003ff88d91004 000
12:36:27 003ffd8a7e518
12:36:27 [   21.489127] User Code: 000003ff88a2f3fa: ec42000c107d        clgij   %r4,16,2
12:36:27 ,000003ff88a2f412
12:36:27 [   21.489127]            000003ff88a2f400: a7440008            brc     4,000003
12:36:27 ff88a2f410
12:36:27 [   21.489127]           #000003ff88a2f404: e70030000837        vll     %v16,%r0
12:36:27 ,0(%r3)
12:36:27 [   21.489127]           >000003ff88a2f40a: e7002000083f        vstl    %v16,%r0
12:36:27 ,0(%r2)
12:36:27 [   21.489127]            000003ff88a2f410: 07fe                bcr     15,%r14
12:36:27 [   21.489127]            000003ff88a2f412: b9040012            lgr     %r1,%r2
12:36:27 [   21.489127]            000003ff88a2f416: ec54ff53a065        clgrj   %r5,%r4,
2:36:27 10,000003ff88a2f2bc
12:36:27 [   21.489127]            000003ff88a2f41c: ec020007ff7d        clgij   %r0,255,
12:36:27 2,000003ff88a2f42a
12:36:27 [   21.490251] Last Breaking-Event-Address:
12:36:27 [   21.490343]  [<000003ff88d89434>] 0x3ff88d89434
12:36:27 [   21.498213] Core dump to |/bin/false pipe failed
12:36:27 [   21.504106] systemd[1]: Caught <ILL>, core dump failed (child 143, code=kille
12:36:27 d, status=4/ILL).
12:36:27 [   21.504246] systemd[1]: Freezing execution.
```

I remembered issue #363 and the problems triggered by Miscellaneous-Instruction-Extensions Facility 2 and 3 support (Facilities 58 and 61 respectively), so tried disabling those facilities. Disabling facility 58 made no difference, but disabling facility 61 prevented the program check.

```
   ::
12:48:18 HHC00898W Facility( 061_MISC_INSTR_EXT_3 ) *Disabled for z/Arch
12:48:18 HHC00007I Previous message from function 'facility_enable_disable' at facility.c(3892)
   ::
12:49:15 
12:49:15 Welcome to [0;38;2;60;110;180mFedora 33 (Thirty Three) dracut-050-64.git20200529
12:49:15 .fc33 (Initramfs)[0m!
12:49:15 
12:49:20 [   21.585441] systemd[1]: /usr/lib/systemd/system/plymouth-start.service:15: Un
12:49:20 it configured to use KillMode=none. This is unsafe, as it disables systemd's pro
   ::
```

The presence of Miscellaneous-Instruction-Extensions Facility 3 is obviously being interpreted as more widespread facilities support!

**Comment by Fish-Git** [comment]
> The presence of Miscellaneous-Instruction-Extensions Facility 3 is obviously being interpreted as more widespread facilities support!

Or at least it is by Fedora s390x, apparently!

So ................. Am I to presume this GitHub Issue was created just to document this fact, or is there something you feel we need, or should, do about it?

Because from my point of view, there's nothing we really _can_ do about it other than to implement support for Vector Facility for z/Architecture (facility bit 129) of course, which would be a _HUGE_ undertaking and one which I myself am not willing to take on right now.

**Comment by mcisho** [comment]
Primarily for documenting. I'll try and work out what's happening, though I don't hold out much hope.

**Comment by Fish-Git** [comment]
> Primarily for documenting.

Then I hope you don't mind if I go ahead and close this issue.

> I'll try and work out what's happening, though I don't hold out much hope.

Understood. If you do figure anything out, feel free to post your findings here. As you know, you can still post comments to closed Issues.

Or you can re-open this Issue again too of course, if you feel doing so is warranted

Thanks, Ian.

**Comment by mcisho** [comment]
The abend was in the **glibc-2.32** package, in the **memmove** function handler.

The memmove function handler has three sequences of instructions, the first for ARCH 13 hardware (i.e. z15 or later), the second for ARCH 11 hardware (i.e. z13 or later), and the third for earlier ARCH hardware. The sequences of instructions for ARCH 11 and 13 hardware both contain vector instructions, either VL/VST or VLL/VSTL pairs, as they operate on (up to) 16-bytes at a time.

The ARCH 13 sequence is called if facility bit 61 is on (i.e. Miscellaneous-Instruction-Extensions Facility 3 is present), the ARCH 11 sequence is called if the hardware supports vector instructions, otherwise the earlier ARCH sequence is called.

Glibc is assuming that facility bit 61 being on indicates the presence of vector instructions.

See glibc-2.32\sysdeps\s390\memmove.c
s390_libc_ifunc_expr (__redirect_memmove, memmove,
          ({
      s390_libc_ifunc_expr_stfle_init ();
      (HAVE_MEMMOVE_ARCH13
       && S390_IS_ARCH13_MIE3 (stfle_bits))
        ? MEMMOVE_ARCH13
        : (HAVE_MEMMOVE_Z13 && (hwcap & HWCAP_S390_VX))
        ? MEMMOVE_Z13
        : MEMMOVE_DEFAULT;
          })
          )
MEMMOVE_ARCH13, MEMMOVE_Z13 and MEMMOVE_DEFAULT are the names of the three sequences of instructions. HAVE_MEMMOVE_ARCH13 and HAVE_MEMMOVE_Z13 are variables initialized at compile time.

**Comment by Fish-Git** [comment]
> Glibc is assuming that facility bit 61 being on indicates the presence of vector instructions.

(sigh) It figures. &nbsp; `:(`

Another case of IBM presuming its architecture can only be run on its own hardware. We are seeing the same lazy programming erroneous presumptions being made with ever increasing frequency in z/OS and z/VM too. &nbsp; `:(`

The truly sad/frustrating thing is, there doesn't really seem to be anything we can do about it. &nbsp; `:(`

Thanks for making the effort to get to the bottom of things, Ian. &nbsp;MUCH appreciated buddy.

**Comment by mcisho** [comment]
It might not be IBM's fault, though lazy programming may have played a part. I'm going to open a bug report with the GNU C Library project, so who knows, some future release of glibc might contain the fix, particularly as I'll be providing the fix in the report.

**Comment by mcisho** [comment]
The GNU C Library project has acknowledged and fixed the bug, and the fix will presumably be included in **glibc-2.34**, which is due to be released on or around **August 1st, 2021**.

**Comment by mcisho** [comment]
The bug has been fixed in **glibc 2.33-16**, which was recently released.
